### PR TITLE
add resource reference e2e test

### PR DESCRIPTION
**Pull request by jaypipes**
Adds an e2e test that verifies resource references (at least from Role -> Policy) are working as expected.

Issue aws-controllers-k8s/community#1610

Signed-off-by: Jay Pipes <jaypipes@gmail.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/tests/test_references.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    assert k8s.get_resource_exists(ref)
+
+    # NOTE(jaypipes): We specifically do NOT wait for the Role to exist in
+    # the IAM API here because we will create the referred-to Policy and
+    # wait for the reference to be resolved
+
+    yield (ref, cr, role_name)
+
+    if k8s.get_resource_exists(ref):
+        # If all goes properly, we should not hit this because the test cleans
```
Comment (jljaco) [inline test/e2e/tests/test_references.py:77]: Nit: why not just `assert not k8s.get_resource_exists(ref)`?

**Review by jljaco — APPROVED** [review]
looks good; 1 nit

File: test/e2e/tests/test_references.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+
+@pytest.fixture(scope="module")
+def referring_role(referred_policy_name):
+    role_name = random_suffix_name("referring-role", 24)
+    role_desc = "a role that refers to a policy"
+
+    replacements = REPLACEMENT_VALUES.copy()
+    replacements['ROLE_NAME'] = role_name
+    replacements['ROLE_DESCRIPTION'] = role_desc
```
Comment (RedbackThomson) [inline test/e2e/tests/test_references.py:51]: lol Do we really need to have this as a parameter? Surely we could have hardcoded this into the YAML

File: test/e2e/tests/test_references.py (hunk 1 above)
Comment (RedbackThomson) [reply test/e2e/tests/test_references.py:77]: This is within the fixture, so we don't want to assert anything after the test has already run (it runs after we `yield` above). This code is just to check that if it wasn't deleted as part of the test, we clean it up now

File: test/e2e/tests/test_references.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            period_length=DELETE_ROLE_TIMEOUT_SECONDS,
+        )
+        assert deleted
+
+        role.wait_until_deleted(role_name)
+
+
+@pytest.fixture(scope="module")
+def referred_policy(referred_policy_name):
+    policy_desc = "a referred-to policy"
```
Comment (RedbackThomson) [inline test/e2e/tests/test_references.py:90]: Critical 👍🏼

File: test/e2e/tests/test_references.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+    yield (ref, cr, policy_arn)
+
+    _, deleted = k8s.delete_custom_resource(
+        ref,
+        period_length=DELETE_POLICY_TIMEOUT_SECONDS,
+    )
+    assert deleted
+
+    policy.wait_until_deleted(policy_arn)
```
Comment (RedbackThomson) [inline test/e2e/tests/test_references.py:128]: Can we do the same conditional for deletion as we have for the other fixture?

File: test/e2e/tests/test_references.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+
+@service_marker
+@pytest.mark.canary
+class TestReferences:
+    def test_references(self, referring_role, referred_policy):
```
Comment (RedbackThomson) [inline test/e2e/tests/test_references.py:134]: Rename to denote as the `role-policy` reference relationship?
```suggestion
def test_policy_role_reference(self, referring_role, referred_policy):
```

File: test/e2e/tests/test_references.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        condition.assert_synced(policy_ref)
+        condition.assert_synced(role_ref)
+
+        role.wait_until_exists(role_name)
+
+        # NOTE(jaypipes): We need to manually delete the Role first because
+        # pytest fixtures will try to clean up the Policy fixture *first*
+        # (because it was initialized after Role) but if we try to delete the
+        # Role before the Policy, the cascading delete protection of resource
+        # references will mean the Role won't be deleted.
```
Comment (RedbackThomson) [inline test/e2e/tests/test_references.py:156]: Won't the reconciler retry the deletion until both are deleted? This seems like a nice edge case for references that we can test

File: test/e2e/tests/test_references.py (hunk 2 above)
Comment (jaypipes) [reply test/e2e/tests/test_references.py:51]: removed unnecessary params.

File: test/e2e/tests/test_references.py (hunk 1 above)
Comment (jaypipes) [reply test/e2e/tests/test_references.py:77]: Yes, what @RedbackThomson said... :)

File: test/e2e/tests/test_references.py (hunk 4 above)
Comment (jaypipes) [reply test/e2e/tests/test_references.py:128]: Well, this one we actually want to delete it in the fixture teardown...

File: test/e2e/tests/test_references.py (hunk 5 above)
Comment (jaypipes) [reply test/e2e/tests/test_references.py:134]: done.

File: test/e2e/tests/test_references.py (hunk 6 above)
Comment (jaypipes) [reply test/e2e/tests/test_references.py:156]: See the explanation... it's because the fixtures are cleaned up in the reverse order of initialization. And so if we just relied on the test case's teardown to delete (both) the role and policy, it would fail the `assert` in the teardown of the Policy fixture because it would not delete the resource until the Role resource was successfully deleted.

Comment (RedbackThomson) [reply test/e2e/tests/test_references.py:151]: Ahhh failing the assertion. Hmm yeah that sucks. Ok

Comment (RedbackThomson) [reply test/e2e/tests/test_references.py:151]: How do you feel about combining both YAML documents into a single file and applying them at the same time? Since we aren't asserting anything between creating the two resources lets just do it simultaneously?
